File: frontend-react/src/api/simulations.ts
import { apiClient } from "./client";

export type SimulationExpense = {
  id: number;
  concept: string;
  amount: number;
};

export type SimulationProject = {
  id: number;
  name: string;
  budget: number;
  subsidyPercent: number;
  thresholdPercent: number;
  expenses: SimulationExpense[];
};

type ApiSimulationExpense = {
  id: number;
  project_id: number;
  concept: string;
  amount: number;
};

type ApiSimulationProject = {
  id: number;
  tenant_id: number | null;
  name: string;
  budget: number;
  subsidy_percent: number;
  threshold_percent: number;
  expenses: ApiSimulationExpense[];
};

export type SimulationProjectCreate = {
  name: string;
  budget?: number;
  subsidyPercent?: number;
  thresholdPercent?: number;
};

export type SimulationProjectUpdate = Partial<SimulationProjectCreate>;

export type SimulationExpenseCreate = {
  concept: string;
  amount?: number;
};

export type SimulationExpenseUpdate = Partial<SimulationExpenseCreate>;

const buildTenantHeaders = (tenantId?: number) =>
  tenantId
    ? {
        headers: {
          "X-Tenant-Id": tenantId.toString(),
        },
      }
    : undefined;

const mapExpense = (expense: ApiSimulationExpense): SimulationExpense => ({
  id: expense.id,
  concept: expense.concept,
  amount: Number(expense.amount ?? 0),
});

const mapProject = (project: ApiSimulationProject): SimulationProject => ({
  id: project.id,
  name: project.name,
  budget: Number(project.budget ?? 0),
  subsidyPercent: Number(project.subsidy_percent ?? 0),
  thresholdPercent: Number(project.threshold_percent ?? 50),
  expenses: project.expenses?.map(mapExpense) ?? [],
});

const serializeProjectPatch = (payload: SimulationProjectUpdate) => ({
  name: payload.name,
  budget: payload.budget,
  subsidy_percent: payload.subsidyPercent,
  threshold_percent: payload.thresholdPercent,
});

export async function fetchSimulations(tenantId?: number) {
  const response = await apiClient.get<ApiSimulationProject[]>(
    "/api/v1/erp/simulations",
    buildTenantHeaders(tenantId),
  );
  return response.data.map(mapProject);
}

export async function createSimulationProject(
  payload: SimulationProjectCreate,
  tenantId?: number,
) {
  const response = await apiClient.post<ApiSimulationProject>(
    "/api/v1/erp/simulations",
    {
      name: payload.name,
      budget: payload.budget ?? 0,
      subsidy_percent: payload.subsidyPercent ?? 0,
      threshold_percent: payload.thresholdPercent ?? 50,
    },
    buildTenantHeaders(tenantId),
  );
  return mapProject(response.data);
}

export async function updateSimulationProject(
  projectId: number,
  payload: SimulationProjectUpdate,
  tenantId?: number,
) {
  const response = await apiClient.patch<ApiSimulationProject>(
    `/api/v1/erp/simulations/${projectId}`,
    serializeProjectPatch(payload),
    buildTenantHeaders(tenantId),
  );
  return mapProject(response.data);
}

export async function deleteSimulationProject(
  projectId: number,
  tenantId?: number,
) {
  await apiClient.delete(
    `/api/v1/erp/simulations/${projectId}`,
    buildTenantHeaders(tenantId),
  );
}

export async function createSimulationExpense(
  projectId: number,
  payload: SimulationExpenseCreate,
  tenantId?: number,
) {
  const response = await apiClient.post<ApiSimulationExpense>(
    `/api/v1/erp/simulations/${projectId}/expenses`,
    {
      concept: payload.concept,
      amount: payload.amount ?? 0,
    },
    buildTenantHeaders(tenantId),
  );
  return mapExpense(response.data);
}

export async function updateSimulationExpense(
  projectId: number,
  expenseId: number,
  payload: SimulationExpenseUpdate,
  tenantId?: number,
) {
  const response = await apiClient.patch<ApiSimulationExpense>(
    `/api/v1/erp/simulations/${projectId}/expenses/${expenseId}`,
    payload,
    buildTenantHeaders(tenantId),
  );
  return mapExpense(response.data);
}

export async function deleteSimulationExpense(
  projectId: number,
  expenseId: number,
  tenantId?: number,
) {
  await apiClient.delete(
    `/api/v1/erp/simulations/${projectId}/expenses/${expenseId}`,
    buildTenantHeaders(tenantId),
  );
}

File: backend-fastapi/app/api/v1/simulations.py
from typing import Optional

from fastapi import APIRouter, Depends, Header, HTTPException, status
from sqlmodel import Session

from app.api.deps import get_current_active_user
from app.db.session import get_session
from app.models.user import User
from app.schemas.simulations import (
    SimulationExpenseCreate,
    SimulationExpenseRead,
    SimulationExpenseUpdate,
    SimulationProjectCreate,
    SimulationProjectRead,
    SimulationProjectUpdate,
)
from app.services.simulations_service import (
    create_simulation_expense,
    create_simulation_project,
    delete_simulation_expense,
    delete_simulation_project,
    list_simulation_projects,
    update_simulation_expense,
    update_simulation_project,
)


router = APIRouter()


def _tenant_for_read(current_user: User, x_tenant_id: Optional[int]) -> Optional[int]:
    # Superadmin puede leer sin tenant; usuarios normales usan su tenant.
    if current_user.is_super_admin:
        return x_tenant_id
    if current_user.tenant_id is None:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Tenant requerido.",
        )
    return current_user.tenant_id


def _tenant_for_write(current_user: User, x_tenant_id: Optional[int]) -> int:
    # Escrituras exigen tenant explicito (superadmin) o tenant del usuario.
    if current_user.is_super_admin:
        if x_tenant_id is None:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="X-Tenant-Id requerido para escribir.",
            )
        return x_tenant_id
    if current_user.tenant_id is None:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Tenant requerido.",
        )
    return current_user.tenant_id


@router.get("/simulations", response_model=list[SimulationProjectRead])
def api_list_simulations(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_active_user),
    x_tenant_id: Optional[int] = Header(default=None, alias="X-Tenant-Id"),
) -> list[SimulationProjectRead]:
    tenant_id = _tenant_for_read(current_user, x_tenant_id)
    return list_simulation_projects(session, tenant_id)


@router.post(
    "/simulations",
    response_model=SimulationProjectRead,
    status_code=status.HTTP_201_CREATED,
)
def api_create_simulation_project(
    payload: SimulationProjectCreate,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_active_user),
    x_tenant_id: Optional[int] = Header(default=None, alias="X-Tenant-Id"),
) -> SimulationProjectRead:
    tenant_id = _tenant_for_write(current_user, x_tenant_id)
    try:
        return create_simulation_project(session, tenant_id, payload)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.patch("/simulations/{project_id}", response_model=SimulationProjectRead)
def api_update_simulation_project(
    project_id: int,
    payload: SimulationProjectUpdate,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_active_user),
    x_tenant_id: Optional[int] = Header(default=None, alias="X-Tenant-Id"),
) -> SimulationProjectRead:
    tenant_id = _tenant_for_write(current_user, x_tenant_id)
    try:
        return update_simulation_project(session, tenant_id, project_id, payload)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc


@router.delete("/simulations/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
def api_delete_simulation_project(
    project_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_active_user),
    x_tenant_id: Optional[int] = Header(default=None, alias="X-Tenant-Id"),
) -> None:
    tenant_id = _tenant_for_write(current_user, x_tenant_id)
    try:
        delete_simulation_project(session, tenant_id, project_id)
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc


@router.post(
    "/simulations/{project_id}/expenses",
    response_model=SimulationExpenseRead,
    status_code=status.HTTP_201_CREATED,
)
def api_create_simulation_expense(
    project_id: int,
    payload: SimulationExpenseCreate,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_active_user),
    x_tenant_id: Optional[int] = Header(default=None, alias="X-Tenant-Id"),
) -> SimulationExpenseRead:
    tenant_id = _tenant_for_write(current_user, x_tenant_id)
    try:
        return create_simulation_expense(session, tenant_id, project_id, payload)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc


@router.patch(
    "/simulations/{project_id}/expenses/{expense_id}",
    response_model=SimulationExpenseRead,
)
def api_update_simulation_expense(
    project_id: int,
    expense_id: int,
    payload: SimulationExpenseUpdate,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_active_user),
    x_tenant_id: Optional[int] = Header(default=None, alias="X-Tenant-Id"),
) -> SimulationExpenseRead:
    tenant_id = _tenant_for_write(current_user, x_tenant_id)
    try:
        return update_simulation_expense(
            session, tenant_id, project_id, expense_id, payload
        )
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc


@router.delete(
    "/simulations/{project_id}/expenses/{expense_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def api_delete_simulation_expense(
    project_id: int,
    expense_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_active_user),
    x_tenant_id: Optional[int] = Header(default=None, alias="X-Tenant-Id"),
) -> None:
    tenant_id = _tenant_for_write(current_user, x_tenant_id)
    try:
        delete_simulation_expense(session, tenant_id, project_id, expense_id)
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
